solutions/platinum/usaco-577.mdx

```diff
@@ -0,0 +1,148 @@
+---
+id: usaco-577
+source: USACO Platinum 2015 December
+title: High Card Low Card
+author: Chongtian Ma
+---
+
+[Official Analysis (C++)](https://usaco.org/current/data/sol_cardgame_platinum_dec15.html)
+
+## Explanation
+
+First, let's determine the cards that bessie will play for each mode.
+Intuitively, if she plays $x$ turns of the "higher mode", then she must play the
+$x$ highest cards in some order. Her score for "higher mode" can't worsen if she
+switches out any card for a lower-numbered card. Even if she does, it might
+worsen her score for the "lower mode". As for her strategy, she always wants to
+play the smallest card available **among all $x$ cards** that is greater than
+Elsie's card. Therefore, the following algorithm is always optimal for "highest
+mode":
+
+```
+Let E = An array of Elsie's first x cards
+let B = An array of Bessie's highest x cards
+score = 0
+for d from 1 to 2n:
+  for elsie_card in E:
+    if bessie_card with number (elsie_card + d) exists in B:
+      remove elsie_card from E
+      remove bessie_card from B
+      add 1 to score
+```
+
+Directly implementing the above algorithm yields quadratic time for each $x$,
+which is too slow. We can use a segment tree to speed things up.
+
+Essentially, the segment tree calculates the score for each $d$ within a
+interval of length $2^k$, where $k$ is the depth of the vertex from its leaves.
+For each vertex of the segment tree, we have to track the following:
+
+- The maximum achievable score after pairing all Bessie Cards with Elsie Cards
+  in the interval.
+- The number of Elsie cards that have not been paired with a Bessie Card.
+- The number of Bessie cards that have not been paired with an Elsie Card.
+
+To merge two intervals into a bigger interval, we can pair all unpaired Elsie
+cards in the left interval with unpaired Bessie cards in the right interval.
+Therefore, the minimum of these two values will contribute to the score for the
+bigger interval.
+
+The algorithm for "lower mode" can be determined similarly.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct segtree {
```

Comment:
(optional)

don't see a well-defined c++ standard for class/struct names, but most of the solution class/struct names here use pascalcase.

Reply:
classes/structs should always use pascalcase here

Reply:
least picky pm